File: lib/src/models/order.dart
import 'package:sneakerx/src/models/custom/address_snapshot.dart';
import 'package:sneakerx/src/models/enums/payment_status.dart';
import 'package:sneakerx/src/models/payment.dart';
import 'package:sneakerx/src/models/shop_order.dart';

class Order {
  final int orderId;
  final int userId;
  final List<Payment> payments;
  final AddressSnapshot shippingAddress;
  final double totalAmount;
  final PaymentStatus paymentStatus;
  final DateTime createdAt;
  final List<ShopOrder> shopOrders;

  Order({
    required this.orderId,
    required this.userId,
    required this.payments,
    required this.shippingAddress,
    required this.totalAmount,
    required this.paymentStatus,
    required this.createdAt,
    required this.shopOrders,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    return Order(
      orderId: json['orderId'],
      userId: json['userId'],
      payments: (json['payments'] as List?)
        ?.map((ele) => Payment.fromJson(ele as Map<String, dynamic>))
        .toList() ?? [],
      shippingAddress: AddressSnapshot.fromJson(json['shippingAddress']),
      totalAmount: json['totalAmount'],
      paymentStatus: PaymentStatus.values.byName(json['paymentStatus']),
      createdAt: DateTime.parse(json['createdAt']),
      shopOrders: (json['shopOrders'] as List?)
        ?.map((ele) => ShopOrder.fromJson(ele as Map<String, dynamic>))
        .toList() ?? [],
    );
  }
}

File: src/main/java/com/example/sneakerx/entities/Order.java
package com.example.sneakerx.entities;

import com.example.sneakerx.entities.customClasses.AddressSnapshot;
import com.example.sneakerx.entities.enums.OrderStatus;
import com.example.sneakerx.entities.enums.PaymentStatus;
import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.type.SqlTypes;

import java.time.LocalDateTime;
import java.util.List;

@Data
@AllArgsConstructor
@NoArgsConstructor
@Entity
@Table(name = "orders")
public class Order {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer orderId;

    @ManyToOne
    @JoinColumn(name = "user_id")
    @JsonIgnore
    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    private User user;

    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Payment> payments;

    @JdbcTypeCode(SqlTypes.JSON) // <--- Tells Hibernate this is a JSON column
    @Column(name = "shipping_address", columnDefinition = "json")
    private AddressSnapshot shippingAddress;

    private Double totalAmount;

    @Enumerated(EnumType.STRING)
    private PaymentStatus paymentStatus;

    @CreationTimestamp
    @Column(updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;

    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<ShopOrder> shopOrders;
}
